### PR TITLE
FIX: get_datalim for collection

### LINKED ISSUE #13639: scatter and date auto-lim far too wide....
### Bug report

```python
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

df = pd.DataFrame({'Depth': [0.2, 0.4, 0.4, 0.4, 0.4, 0.4, 0.6, 0.4, 3.2, 2.0],
                   'DateTimeUTC': [pd.Timestamp('2018-03-28 06:25:08'),
                                   pd.Timestamp('2018-03-28 06:25:49'),
                                   pd.Timestamp('2018-03-28 06:27:06'),
                                   pd.Timestamp('2018-03-28 06:32:11'),
                                   pd.Timestamp('2018-03-28 06:32:59'),
                                   pd.Timestamp('2018-03-28 06:34:02'),
                                   pd.Timestamp('2018-03-28 06:35:38'),
                                   pd.Timestamp('2018-03-28 06:37:04'),
                                   pd.Timestamp('2018-03-28 06:39:08'),
                                   pd.Timestamp('2018-03-28 06:40:52')]})

x = np.asarray(df['DateTimeUTC'])
y = np.asarray(df['Depth'])

fig, axs = plt.subplots(2, 1)


ax = axs[0]
ax.scatter(x, y)
ax.set_title("using scatter")

ax = axs[1]
ax.plot(x, y)
ax.set_title("Using plot")

fig.tight_layout()
plt.show()
```

Figure_1

**Matplotlib version**
  * Matplotlib version: master

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

EDIT 12 Mar:

Closes #13639

Closes the non-unique auto-limit problem in #7413 for collections.  Collections have "offsets" and "paths" and each can be in a different co-ordinate frames;  This PR tackles the autolim problem by consulting the `self.get_transform()` and `self.get_offset_transform()` for the collection and checking if these are `ax.dataTrans` or not.  

For `collection.get_datalim`: There are two cases that I can see in common usage: 

A. `offsets=None` for `LineCollection`.  In this case, all the vertices in paths should be included in the autolim if `transform=ax.dataTrans`. (this is the case for `streamplot`)

B `offset` is not None: 

  1. if the offset coords (`offsetTrans`) is not the axes `dataTrans`, then don't auto-lim on this collection at all.  I don't know off hand of any built-in examples like this, but if the user specifies a bunch of markers in AxesTrans co-ords, then they shouldn't expect the axes to change its x/ylims.  

  2. If the path coords (`self.transform`) is `dataTrans` then we use the existing algorithm.  This is useful for paths like in `quiver`, where the offset defines the tail of an arrow, and the path defines it's length.  In that case, the whole arrow is meant to be in data space, and the lims can change size accordingly 

  3. if the path co-ords are *not* `dataTrans`, then we don't consult the `path` at all, because these paths are in some other unit that doesn't help set the data limits.  i.e. scatter markers are in points, and these can be arbitrarily large (even larger than the figure if you want), so setting the axes *data* limits based on this size is ill-defined.  That lead to many of the issues in #7413. Of course the `offset` limits should still be included in the limits (i,e. we want the centre of markers to be in the new x/ylims.)

This necessitated 14 image tests needing to be changed, *mostly* due to small axes limit differences in `scatter` tests.  

The major exception is `test_transforms.py::test_pre_transform_plotting` which highlights a limitation of the consult-the-transforms approach used here.  If the user passes a simple transform like `transform = mtransforms.Affine2D().scale(10) + ax.dataTrans` the autolimiting will not take place because `transform != ax.dataTrans`.  Thats a cost of the approach, but I *personally* feel users should just transform the data before asking matplotlib to plot it, or be willing to manually adjust the axes limits.  

```python 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

df = pd.DataFrame({'Depth': [0.2, 0.4, 0.4, 0.4, 0.4, 0.4, 0.6, 0.4, 3.2, 2.0],
                   'DateTimeUTC': [pd.Timestamp('2018-03-28 06:25:08'),
                                   pd.Timestamp('2018-03-28 06:25:49'),
                                   pd.Timestamp('2018-03-28 06:27:06'),
                                   pd.Timestamp('2018-03-28 06:32:11'),
                                   pd.Timestamp('2018-03-28 06:32:59'),
                                   pd.Timestamp('2018-03-28 06:34:02'),
                                   pd.Timestamp('2018-03-28 06:35:38'),
                                   pd.Timestamp('2018-03-28 06:37:04'),
                                   pd.Timestamp('2018-03-28 06:39:08'),
                                   pd.Timestamp('2018-03-28 06:40:52')]})

x = np.asarray(df['DateTimeUTC'])
y = np.asarray(df['Depth'])

fig, axs = plt.subplots(2, 1)

ax = axs[0]
#pp = ax.plot(x, y, '.')
ax.scatter(x, y, s=25, marker='d')
ax.set_title("using scatter")

ax = axs[1]
ax.plot(x, y)
ax.set_title("Using plot")

fig.tight_layout()
plt.show()

```

### Before

Figure_2

### After

Figure_1

## Big Markers 

Note that big markers will be clipped.  This is "worse", but its consistent... 

### Before:

boo2

### After:

boo

## Cures inconsistency 
To see the inconsistency in the old way not how the xlim changes with each re-lim:

```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots()

r = 2 * 72
sc = ax.scatter([0, 1], [0, 1], s=r * r * 3.14)

plt.show()
xmin = []
for i in range(4):
    fig.canvas.draw()
    ax.update_datalim(sc.get_datalim(ax.transData))
    ax.autoscale()
    print(ax.get_xlim())
    xmin += [ax.get_xlim()[0]]
```

```
(-0.5934780784674604, 1.59347807846746)
(-0.9094406858496092, 1.9094406858496091)
(-1.15777773561647, 2.15777773561647)
(-1.3529631431489448, 2.3529631431489455)
```

Of course the resulting markers are still clipped anyways:

### Before:
boo3

### After

```
(-0.05, 1.05)
(-0.05, 1.05)
(-0.05, 1.05)
(-0.05, 1.05)
```

boo4

#### `test_axes`

Failures are minor axes limits changing slightly....

#### `test_collections.py`

`test_barb_limits` fails because it is constructed as: 

```python
    x = np.linspace(-5, 10, 20)
    y = np.linspace(-2, 4, 10)
    y, x = np.meshgrid(y, x)
    trans = mtransforms.Affine2D().translate(25, 32) + ax.transData
    plt.barbs(x, y, np.sin(x), np.cos(y), transform=trans)
```

and that fails the rubric above because the offset transform is not the data transform.  Not sure about how people feel about this - personally I'm not a fan of defining data at one location and then using the  transform stack to move the object around, instead of just manipulating the original data, and if folks want to do this, I'm OK with auto_lims failing for them.  If you are going to manually move things around with a low-level stack, then you can also be expected to set your limits accordingly: flexibility comes at the cost of automation.  

`test_EllipseCollection`, `test_regularpolycollection_rotate` fail somewhat badly because they are big paths that used to get some space from the existing algorithm, and no longer do.  `scatter_post_alpha` is a very minor change....

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
This is WIP from a still-need-to-fix-tests point of view, but I wanted to have a concrete proposal on the table for fixing part of #7413.

**Comment by anntzer** [comment]
I think the general approach is good, haven't reviewed carefully yet.
Does this close #6915/#11198 too?
In https://github.com/matplotlib/matplotlib/issues/7413#issuecomment-260136914 @tacaswell argued in favor of accepting clipped markers too.  Despite my comments on that thread I now also think ignoring marker extents in scatter makes complete sense (the "larger-than-figure marker" case is quite convincing).

**Comment by jklymak** [comment]
No it doesn’t fix #6915. But I don’t think it’s incompatible with your PR leaving the autolim to draw time which makes sense to me.

**Comment by jklymak** [comment]
OK, this breaks `streamplots`, mostly in manageable ways, but a couple of the failures I can't figure out yet.

File: lib/matplotlib/collections.py [hunk 1]
```diff
@@ -181,7 +181,12 @@ def get_offset_transform(self):
     def get_datalim(self, transData):
         transform = self.get_transform()
         transOffset = self.get_offset_transform()
+        if not transOffset == transData:
```
Comment (anntzer) [inline lib/matplotlib/collections.py]: transOffset != transData

**Comment by jklymak** [comment]
So, this PR damages autolim-ing on constructs like in `test_transforms`:

```python
times10 = mtransforms.Affine2D().scale(10)
ax.scatter(np.linspace(0, 10), np.linspace(10, 0), transform=times10 + ax.transData)
```

I'm OK with that, but wanted to let folks know... The new test file looks fine, but the limits have changed substantially:  https://github.com/matplotlib/matplotlib/blob/d3f8447cc69b21be885c8683c942043e2903ac7f/lib/matplotlib/tests/baseline_images/test_transforms/pre_transform_data.png

**Comment by jklymak** [comment]
BTW, is there an easy way to tell if a transform has `ax.dataTrans` in it?  i.e. if someone has done a simple `offsetTrans+ax.dataTrans`?

**Comment by anntzer** [comment]
@jklymak Have a look at contains_branch or contains_branch_separately (not exactly sure about their use, but I *think* they may be what you want).

**Comment by jklymak** [comment]
Please see updated description above.  

I can make an API change doc etc, but wanted feedback on the approach first.  Ping @ImportanceOfBeingErnest as you have been known to make extensive use of the transform machinery.  This will break some autolimits for some transforms.

**Comment by anntzer** [comment]
Instead of changing the baseline images, can you keep the old ones, add an assert that the x/ylims are what they should be per the new approach, then force the x/ylims to match the old values?

**Comment by anntzer** [comment]
I'll wait until you look into contains_branch and possibly don't change the baseline images?
But again, generally speaking I like the approach.

**Comment by jklymak** [comment]
> Instead of changing the baseline images, can you keep the old ones, add an assert that the x/ylims are what they should be per the new approach, then force the x/ylims to match the old values?

Moved this discussion to gitter :wink:

**Comment by jklymak** [comment]
> I'll wait until you look into contains_branch...

Seems to work!  The `transform=times10 + ax.transData` now works, though the image still changes because the scatter in that test changes its limits....

**Comment by anntzer** [comment]
copypasting my reply on gitter:
> Instead of regenerating the images, it may also be worth just checking whether we could do away with some of the tests, in case they're redundant with others
For example the test that generates scatter.png could become a check_figures_equal comparing scatter() with manually calling plot() to add one marker at a time
scatter_post_alpha too
I guess colorbar_single_scatter could just not do an image check but check that the colorbar bounds are correctly set
Also, fewer baseline-image tests also help mplcairo :) (as they are renderer-dependent)

**Comment by jklymak** [comment]
Fair enough - I'll look at rejiggering the tests after we get some consensus that this is a good way to go.  Particularly as A/B-ing the tests shows the end-effects of this change for users....

**Comment by jklymak** [comment]
...alos ping @efiring

File: lib/matplotlib/tests/test_streamplot.py [hunk 2]
```diff
@@ -92,11 +92,11 @@ def test_direction():
...
 def test_streamplot_limits():
     ax = plt.axes()
-    x = np.linspace(-5, 10, 20)
-    y = np.linspace(-2, 4, 10)
+    x = np.linspace(-5, 10, 20) + 25
+    y = np.linspace(-2, 4, 10) + 32
     y, x = np.meshgrid(y, x)
-    trans = mtransforms.Affine2D().translate(25, 32) + ax.transData
-    plt.barbs(x, y, np.sin(x), np.cos(y), transform=trans)
+#    trans = mtransforms.Affine2D().translate(25, 32) + ax.transData
```
Comment (anntzer) [inline lib/matplotlib/tests/test_streamplot.py]: just delete this?

File: lib/matplotlib/collections.py [hunk 3]
```diff
@@ -179,9 +181,30 @@ def get_offset_transform(self):
         return t
 
     def get_datalim(self, transData):
+
+        # Get the automatic datalim of the collection.
+        #
+        # This operation depends on the transforms for the data in the
+        # collection and whether the collection has offsets.
+        #
```
Comment (anntzer) [inline lib/matplotlib/collections.py]: comment needs update re: contains_branch

File: lib/matplotlib/tests/test_streamplot.py (hunk 2 above)
Comment (jklymak) [reply lib/matplotlib/tests/test_streamplot.py]: Oops, meant to re-Instate this actually.  Thanks....

File: lib/matplotlib/collections.py [hunk 4]
```diff
@@ -179,9 +181,30 @@ def get_offset_transform(self):
...
+        #    a) transform = transData: use the path + offset for limits (i.e
+        #    for bar).
+        #    b) transform != transData: just use the offsets for the limits
+        #    (i.e. for scatter)
+        # 3) otherwise return a null Bbox.
+
         transform = self.get_transform()
         transOffset = self.get_offset_transform()
+        if (not self._offsetsNone and
+            not transOffset.contains_branch(transData)):
```
Comment (anntzer) [inline lib/matplotlib/collections.py]: I guess(?) you could get a few extra points by checking `contains_branch_separately` and then deciding to use the `x` or the `y` data limits if one of them does contain transData; this could be a separate PR too as the use case is a bit obscure.

**Comment by anntzer** [comment]
btw, I don't mind doing the conversion to check_figures_equal if you want.

**Comment by ImportanceOfBeingErnest** [comment]
I will need to look at this more thoroughly. Based on a quick glance I wonder if we can't keep using `mpath.get_path_collection_extents` even in the cases you exclude here. It does not seem to give generally bad output, it just needs a good frozen transform to start with (maybe that is the hard part, but maybe not?) and it will then not be 100% accurate - but probably still better then cropping half the marker.

File: lib/matplotlib/collections.py [hunk 5]
```diff
@@ -195,13 +218,32 @@ def get_datalim(self, transData):
             offsets = offsets.filled(np.nan)
             # get_path_collection_extents handles nan but not masked arrays
 
+        result = transforms.Bbox.null()
```
Comment (anntzer) [inline lib/matplotlib/collections.py]: I guess the conditionals below are more clearly written as
```python
# (don't define result here)
if len(paths) and len(offsets) and not self._offsetsNone:
    if transforms.contains_branch(transData):
        # one case.
        return ...
    else:
        # other case.
        return ...
else:
    return transforms.Bbox.null()
```

**Comment by jklymak** [comment]
> I will need to look at this more thoroughly. Based on a quick glance I wonder if we can't keep using `mpath.get_path_collection_extents` even in the cases you exclude here. It does not seem to give generally bad output, it just needs a good frozen transform to start with (maybe that is the hard part, but maybe not?) and it will then not be 100% accurate - but probably still better then cropping half the marker.

We could heuristically do that when the marker size is not too big (probably a quarter of the axes size), but you still need to solve the inverse problem, which isn't necessarily trivial particularly if you need to take into account the scale of the axes, or iterate to a solution as discussed before.

**Comment by jklymak** [comment]
One further point, this doesn't typically cut off markers because of the default data margins.  Just really big markers...  Which are also the ones that cause auto-lim instability...

File: lib/matplotlib/collections.py (hunk 5 above)
Comment (jklymak) [reply lib/matplotlib/collections.py]: Well, "other case" has to run if `self._offsetsNone` is `True` which it won't in the above...

Comment (anntzer) [reply lib/matplotlib/collections.py]: then perhaps
```python
if len(paths) and len(offsets):
    if contains(transData):
        # other case
        return ...
    if not self._offsetsNone:
        # first case
        return ...
return null()
```
?

**Comment by anntzer** [comment]
Let's try to get some variant of this in in the 3.2 time frame.

**Comment by jklymak** [comment]
> Let's try to get this (or some variant of it) in in the 3.2 time frame.

For sure - unfortunately I've not been able to make the weekly meeting for a while,*but* I think I can make it this coming week and we can/should discuss?

**Comment by tacaswell** [comment]
I am :+1: on only setting the limits based on the location data (not the marker size).

File: lib/matplotlib/collections.py [hunk 6]
```diff
@@ -196,13 +219,30 @@ def get_datalim(self, transData):
...
+                # (i.e. like scatter). We can't uniquely set limits based on
+                # those shapes, so we just set the limits based on their
+                # location.
+                # Finish the transform:
+                offsets = transOffset.transform(offsets)
+                offsets = transData.inverted().transform(offsets)
+                offsets = np.ma.masked_invalid(offsets)
+                if not offsets.mask.all():
+                    extents = [[np.min(offsets[:, 0]), np.min(offsets[:, 1])],
+                               [np.max(offsets[:, 0]), np.max(offsets[:, 1])]]
```
[the remark this answers is not in the document]
Comment (jklymak) [reply lib/matplotlib/collections.py]: Thanks!  Sorry it took a while to get back to this.

**Comment by jklymak** [comment]
@eefiring,  no tests images were removed, though I agree that git makes it look like `lib/matplotlib/tests/baseline_images/test_transforms/pre_transform_data.pdf` was removed.  But its not gone, its still there. 🤷‍♂

File: doc/api/next_api_changes/2019-03-04-AL.rst [hunk 7]
```diff
@@ -13,6 +13,41 @@ without an explicit call to `Axes.autoscale_view`.
 In some cases, this can result in different limits being reported.  If this is
 an issue, consider triggering a draw with `fig.canvas.draw`.
 
+Autoscaling changes for Collections
+```````````````````````````````````
+
+Autoscaling has also changed for artists that are based on the `.Collections`
```
Comment (QuLogic) [inline doc/api/next_api_changes/2019-03-04-AL.rst]: The class is `Collection` (singular), here and below.
```suggestion
Autoscaling has also changed for artists that are based on the `.Collection`
```

File: doc/api/next_api_changes/2019-03-04-AL.rst [hunk 8]
```diff
@@ -13,6 +13,41 @@ without an explicit call to `Axes.autoscale_view`.
...
+class.  Previously, the method that calculates the automatic limits
+`.Collections.get_datalim` tried to take into account the size of objects
```
Comment (QuLogic) [inline doc/api/next_api_changes/2019-03-04-AL.rst]:
```suggestion
`.Collection.get_datalim` tried to take into account the size of objects
```

File: doc/api/next_api_changes/2019-03-04-AL.rst [hunk 9]
```diff
@@ -13,6 +13,41 @@ without an explicit call to `Axes.autoscale_view`.
...
+in the collection and make the limits large enough to not clip any of the
+object.  i.e. for `.Axes.scatter` it would make the limits large enough to not
```
Comment (QuLogic) [inline doc/api/next_api_changes/2019-03-04-AL.rst]:
```suggestion
object, i.e., for `.Axes.scatter` it would make the limits large enough to not
```

File: doc/api/next_api_changes/2019-03-04-AL.rst [hunk 10]
```diff
@@ -13,6 +13,41 @@ without an explicit call to `Axes.autoscale_view`.
...
+specified in physical space, or figure-relative space, because the transform
+from physical units to data limits requires knowing the data limits, and
+becomes invalid when the new limits are applied.  This is an inverse
+problem that is theoretically solvable (if the object is physically smaller
+than the axes), but the extra complexity was not deemed worth it, particularly
+as the most common use case is for markers in scatter that are usually small
+enough to be accommodated by the default data limit margins.
+
+While the new behavior is algorithmically simpler, it is conditional on
+properties of the `.Collections` object:
```
Comment (QuLogic) [inline doc/api/next_api_changes/2019-03-04-AL.rst]:
```suggestion
properties of the `.Collection` object:
```

File: doc/api/next_api_changes/2019-03-04-AL.rst [hunk 11]
```diff
@@ -13,6 +13,41 @@ without an explicit call to `Axes.autoscale_view`.
...
+
+  1. ``offsets = None``, ``transform`` is a child of `.transData`: use the paths
+     for the automatic limits (i.e. for `.LineCollection` in `Axes.streamline`).
```
Comment (QuLogic) [inline doc/api/next_api_changes/2019-03-04-AL.rst]: Doesn't link; is this `Axes.streamplot`?

File: doc/api/next_api_changes/2019-03-04-AL.rst [hunk 12]
```diff
@@ -13,6 +13,41 @@ without an explicit call to `Axes.autoscale_view`.
...
+  2.  ``offsets != None``: and ``offset_transform`` is child of `transData`:
```
Comment (QuLogic) [inline doc/api/next_api_changes/2019-03-04-AL.rst]:
```suggestion
  2.  ``offsets != None``, and ``offset_transform`` is child of `.Axes.transData`:
```

File: doc/api/next_api_changes/2019-03-04-AL.rst [hunk 13]
```diff
@@ -13,6 +13,41 @@ without an explicit call to `Axes.autoscale_view`.
...
+
+    a) ``transform` is child of `.transData`: use the ``path + offset`` for
```
Comment (QuLogic) [inline doc/api/next_api_changes/2019-03-04-AL.rst]:
```suggestion
    a) ``transform`` is child of `.Axes.transData`: use the ``path + offset`` for
```

File: doc/api/next_api_changes/2019-03-04-AL.rst [hunk 14]
```diff
@@ -13,6 +13,41 @@ without an explicit call to `Axes.autoscale_view`.
...
+becomes invalid when the new limits are applied.  This is an inverse
+problem that is theoretically solvable (if the object is physically smaller
+than the axes), but the extra complexity was not deemed worth it, particularly
+as the most common use case is for markers in scatter that are usually small
+enough to be accommodated by the default data limit margins.
+
+While the new behavior is algorithmically simpler, it is conditional on
+properties of the `.Collections` object:
+
+  1. ``offsets = None``, ``transform`` is a child of `.transData`: use the paths
```
Comment (QuLogic) [inline doc/api/next_api_changes/2019-03-04-AL.rst]: Doesn't link; probably clearer to mention the class, too:
```suggestion
  1. ``offsets = None``, ``transform`` is a child of `.Axes.transData`: use the paths
```
(I'm not sure if it will link with the class.)

File: doc/api/next_api_changes/2019-03-04-AL.rst [hunk 15]
```diff
@@ -13,6 +13,41 @@ without an explicit call to `Axes.autoscale_view`.
...
+     for the automatic limits (i.e. for `.LineCollection` in `Axes.streamline`).
+  2.  ``offsets != None``: and ``offset_transform`` is child of `transData`:
+
+    a) ``transform` is child of `.transData`: use the ``path + offset`` for
+        limits (i.e for `Axes.bar`).
```
Comment (QuLogic) [inline doc/api/next_api_changes/2019-03-04-AL.rst]:
```suggestion
        limits (i.e., for `.Axes.bar`).
```

File: doc/api/next_api_changes/2019-03-04-AL.rst [hunk 16]
```diff
@@ -13,6 +13,41 @@ without an explicit call to `Axes.autoscale_view`.
...
+    b) ``transform`` is not a child of `transData`: just use the offsets
```
Comment (QuLogic) [inline doc/api/next_api_changes/2019-03-04-AL.rst]:
```suggestion
    b) ``transform`` is not a child of `.Axes.transData`: just use the offsets
```

File: doc/api/next_api_changes/2019-03-04-AL.rst (hunk 11 above)
Comment (jklymak) [reply doc/api/next_api_changes/2019-03-04-AL.rst]: Thanks @QuLogic for checking the links!

File: lib/matplotlib/collections.py [hunk 17]
```diff
@@ -196,13 +219,30 @@ def get_datalim(self, transData):
...
+                    transform.frozen(), paths, self.get_transforms(),
+                    offsets, transOffset.frozen())
+                return result.inverse_transformed(transData)
+            if not self._offsetsNone:
+                # this is for collections that have their paths (shapes)
+                # in physical, axes-relative, or figure-relative units
+                # (i.e. like scatter). We can't uniquely set limits based on
+                # those shapes, so we just set the limits based on their
+                # location.
+                # Finish the transform:
```
Comment (anntzer) [inline lib/matplotlib/collections.py]: in the same spirit as https://github.com/matplotlib/matplotlib/pull/14845 this could be
```
offset_to_data = transOffset - transData
offsets = offset_to_data.transform(offsets)
```
(you also save a transform by doing so).

Comment (jklymak) [reply lib/matplotlib/collections.py]: I wasn't paying attention when #14845 went in.  This requires remembering that transA - transB is the same as what I wrote here.  I don't think it helps the readability/comprehensibility of all the transform magic to have arithmetic operators that do stuff versus just using explicit methods.

Comment (anntzer) [reply lib/matplotlib/collections.py]: The main operator is trA+trB, which just means "perform trA, then perform trB"; once you know it it's quite natural that trA-trB means "perform trA, then perform the inverse of trB".  You could write `mtransforms.composite_transform_factory(trA, trB.inverse())` if you prefer explicit method names, but the point is that it is more efficient to first multiply two small (3x3) matrices and then use the result to transform a big (N,2) array, rather than to do two transformations in a series.

Comment (jklymak) [reply lib/matplotlib/collections.py]: Good point.  Split the difference:  `(trA + trB.inverted())` seems straightforward enough.  Like, I understand the `__sub__` convention, I just find it a bit obscure.  The `__add__` convention is fine.

File: lib/matplotlib/collections.py (hunk 4 above)
Comment (anntzer) [reply lib/matplotlib/collections.py]: let's just forget this case for now

Comment (jklymak) [reply lib/matplotlib/collections.py]: As discussed on gitter, I think thats best for simplicity sake.  For instance, I don't know what the datalim handling does with infs (for instance) and adding this case is a good chunk of code.  As also discussed, very happy to revisit if it turns out to be needed.

**Comment by jklymak** [comment]
Thanks for the help and suggestions @anntzer  and @efiring !

File: doc/api/next_api_changes/2019-03-04-AL.rst [hunk 18]
```diff
@@ -13,6 +13,41 @@ without an explicit call to `Axes.autoscale_view`.
...
+problem that is theoretically solvable (if the object is physically smaller
+than the axes), but the extra complexity was not deemed worth it, particularly
+as the most common use case is for markers in scatter that are usually small
+enough to be accommodated by the default data limit margins.
+
+While the new behavior is algorithmically simpler, it is conditional on
+properties of the `.Collection` object:
+
+  1. ``offsets = None``, ``transform`` is a child of `.Axes.transData`: use the paths
+     for the automatic limits (i.e. for `.LineCollection` in `Axes.streamplot`).
```
Comment (QuLogic) [inline doc/api/next_api_changes/2019-03-04-AL.rst]: One link fix:
```suggestion
     for the automatic limits (i.e. for `.LineCollection` in `.Axes.streamplot`).
```
